Background job  for long running reports.

shriharishastry opened the pull request:
Pull-Request
 User can mark a Report as Background Report, this would not trigger reports when user tries to load report instance. Instead user can select filter and click Run in Background, if process is initiated a pop with a message and a link to track the report status. 

User can check a report saying it is of type background.

![save_report](https://user-images.githubusercontent.com/10257870/41596586-6a25d694-73e8-11e8-8c48-d2acca25daa4.png)


When user opens report instance , he can set filters if there are any and trigger a button to run in background.
![preview_report](https://user-images.githubusercontent.com/10257870/41596603-78847cfe-73e8-11e8-9a84-149e21519331.png)

A popup(modal) would say where to access results.
![modal](https://user-images.githubusercontent.com/10257870/41596619-8a6d8ad2-73e8-11e8-887f-62a3d2480b2b.png)

User can access  and track these results.
![result_list](https://user-images.githubusercontent.com/10257870/41596842-3db237b4-73e9-11e8-84c6-e8130bc4df8b.png)

Also user can put normal reports to background also, so he would have a saved version on server.

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you lint your code locally prior to submission?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
- [x] Have you successfully run tests with your changes locally?
- [ ] Does your commit message have an explanation for your changes and why you'd like us to include them?
- [ ] Docs have been added / updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Did you modify the existing test cases? If yes, why?

---

What type of a PR is this? 

- [ ] Changes to Existing Features
- [x] New Feature Submissions
- [ ] Bug Fix
- [ ] Breaking Change

--- 

- Motivation and Context (What existing problem does the pull request solve):
- Related Issue: 
[Issue](https://github.com/frappe/frappe/issues/5667)
- Screenshots (if applicable, remember, a picture tells a thousand words):

shridarpatil committed:
File upload without socketio

System setting to select socketio for upload

Commit:
initial changes on query report

Commit:
initial changes on query report

Commit:
initial changes on query report

Commit:
initial setup for background report

Commit:
initial setup for background report

Commit:
initial setup for background report

Commit:
initial setup for background report

Commit:
initial setup for background report

Commit:
initial setup for background report

Commit:
changed arguments for background run

shriharishastry committed:
changed arguments for background run

Commit:
changed arguments for background run

shriharishastry committed:
changed arguments for background run

Commit:
changed arguments for background run

shriharishastry committed:
changed arguments for background run

shriharishastry committed:
added comment

shriharishastry committed:
added comment

shriharishastry committed:
added comment

Commit:
Merge branch 'background-reports' of https://github.com/shriharishastry/frappe into background-reports

shriharishastry committed:
deleted logger file

shriharishastry committed:
deleted logger file

shriharishastry committed:
deleted logger file

shriharishastry committed:
deleted logger file

shriharishastry committed:
code refactor based on review

shriharishastry committed:
code refactor based on review

shriharishastry committed:
code refactor based on review

shriharishastry committed:
code refactor based on review

shriharishastry committed:
removed file and added stringIO

shriharishastry committed:
removed file and added stringIO

shriharishastry committed:
removed file and added stringIO

shriharishastry committed:
removed file and added stringIO

shriharishastry committed:
added button markup and respective js changes

shriharishastry committed:
added button markup and respective js changes

shriharishastry committed:
added button markup and respective js changes

shriharishastry committed:
added button markup and respective js changes

shriharishastry committed:
changes on fields and test copy

shriharishastry committed:
changes on fields and test copy

shriharishastry committed:
changes on fields and test copy

shriharishastry committed:
changes on fields and test copy

Commit:
added redirect link

shriharishastry committed:
added redirect link

Commit:
added redirect link

Commit:
added redirect link

Commit:
added redirect link

Commit:
Merge branch 'background-reports' of https://github.com/shriharishastry/frappe into background-reports

shriharishastry committed:
changes on field naming and redirect to form

shriharishastry committed:
changes on field naming and redirect to form

shriharishastry committed:
changes on field naming and redirect to form

shriharishastry committed:
changes on field naming and redirect to form

shriharishastry committed:
changes on doctype naming

shriharishastry committed:
changes on doctype naming

shriharishastry committed:
changes on doctype naming

shriharishastry committed:
changes on doctype naming

shriharishastry committed:
changes for csv headers

shriharishastry committed:
changes for csv headers

shriharishastry committed:
changes for csv headers

shriharishastry committed:
changes for csv headers

shriharishastry committed:
changes for variable name

shriharishastry committed:
changes for variable name

shriharishastry committed:
changes for variable name

shriharishastry committed:
changes for variable name

shriharishastry committed:
removed whitelist and version number

shriharishastry committed:
removed whitelist and version number

shridarpatil committed:
if not user get user from session

made filters optional

shridarpatil committed:
if not user get user from session

made filters optional

shridarpatil committed:
if not user get user from session

made filters optional

shriharishastry committed:
unit tests

shriharishastry committed:
unit tests

Commit:
Merge branch 'background-reports' of https://github.com/zerodhatech/frappe into background-reports

Commit:
merge issues with remote branch

shriharishastry committed:
merge issues with remote branch

shriharishastry committed:
merge issues with remote branch

shriharishastry committed:
test changes

shriharishastry committed:
test changes

shriharishastry committed:
test changes

shriharishastry committed:
test changes

Commit:
test for completion

Commit:
test for completion

shriharishastry committed:
changes for test

shriharishastry committed:
changes for test

Commit:
removed foreground process

Commit:
removed foreground process

shridarpatil committed:
Merge pull request #2 from shriharishastry/background-reports

Background reports

shriharishastry committed:
changes on unit tests

shriharishastry committed:
changes on unit tests

realvinay committed:
Merge pull request #3 from shriharishastry/background-reports

changes on unit tests

Commit:
indent and stringIo

Commit:
indent and stringIo

shriharishastry committed:
unit tests changes

shriharishastry committed:
unit tests changes

Commit:
Merge branch 'background_report' of https://github.com/zerodhatech/frappe into background-reports

Commit:
tab to spaces

Commit:
tab to spaces

Commit:
tab to spaces

Commit:
tab to spaces

Commit:
tab to spaces

Commit:
tab to spaces

Commit:
unit test

Commit:
unit test

rmehta commented on frappe/core/doctype/background_report_result/background_report_result.py:
```diff
@@ -0,0 +1,63 @@
+
+
+def create_csv_file(columns, data, dt, dn):
+	# create the list of column labels
```
see `frappe.utils.csvutils`

rmehta commented on frappe/public/js/frappe/views/reports/query_report.js:
```diff
@@ -53,6 +53,11 @@ frappe.views.QueryReport = class QueryReport extends frappe.views.BaseList {
 		};
+		
+		this.report_action = {
+			label: __('Run In Background'),
```
this is not very user friendly. The decision to run automatically should be made by the system based on certain "best guesses"

Edit 1: Or in query report, we can specify if this report is to be run in the background only. This seems to be a better decision.

Edit 2: In the query report, we already tag the reference doctype, so we can run a `count` query on the reference table and if its bigger than a number, push it to background.

realvinay replied:
Hi Rushab,

For Edit 1: A checkbox has already been added in report doctype to mark reports to be run in background mode by default. This 'Run In Background' was added to have a submit form button on report view, so that form submission happens after all filters are set by user. But yes, this should not show up for all reports and should only be shown on reports with the background flag set in report definition. 

Also is there a better way to prevent report from triggering before user has entered all the filters ?

For Edit-2 : This functionality has to be available to all types for reports (Query, Script, Standard), so count query might not be feasible. Also the whole intention of building this is to run reports that take several minutes, hours and sometimes days also :). So, a count query is also very expensive on these tables most of the time and also the report might not be just dependent on the table at all.

rmehta replied:
Cool, then lets remove this option. If its a prepared report, then just show download button for the last report (with the relevant filters) with the timestamp or show status of the current one under preparation.

Count is not required if its a prepared report.

For filters, you can add a validation on the server side to make sure that the report is only triggered with filters.

(maybe we should have filters as a child table in the report, so we can control them more effectively)

rmehta commented on frappe/public/js/frappe/views/reports/query_report.js:
```diff
@@ -217,18 +235,39 @@ frappe.views.QueryReport = class QueryReport extends frappe.views.BaseList {
+				const data = r.message;
+				this.toggle_nothing_to_show(true);
+				frappe
+				.msgprint("Background job initiated successfully. Track and access results  <a class='text-info' target='_blank' href="+data.redirect_url+">here</a>",
```
Here, we can show a modal with the last "Prepared Report" and timestamp, so user can download it without opening a new doctype.

rmehta commented on frappe/core/doctype/background_report_result/background_report_result.json:
```diff
@@ -0,0 +1,308 @@
+{
```
How about "Prepared Report"

rmehta reviewed (changes requested):
Thanks for starting this 👍 

See my comments!

rmehta commented:
Summarizing:

- [ ] In Query Report add check `Run in Background`
- [ ] Before running, run a `COUNT` query to check the size of the reference table (there is already a `frappe.db.get_count` that also uses caches 😉 )
- [ ] In the view, let the user download the last prepared report or queue a new one. Make sure there is only one report queued of a particular type at a time.

Commit:
io of byte

Commit:
io of byte

pratu16x7 committed:
[global-search] trim varchar columns

pratu16x7 committed:
[global-search] trim strings directly

pratu16x7 committed:
[global-search] route check

pratu16x7 committed:
Merge pull request #5709 from pratu16x7/hotfix

[global-search] trim varchar columns

saurabh6790 committed:
[security][fix] Sanitize search fields to avoid sql injection (#5713)

* [security][fix] Sanitize search fields to avoid sql injection

* Test Cases for Sanitizer

* Test Cases fix

* [fix] test case

rmehta committed:
[fix] varchar_len in global_search.py

Commit:
prepared report

Commit:
prepared report

Commit:
prepared report

Commit:
prepared report

Commit:
test cases

Commit:
test cases

Commit:
test cases

Commit:
test cases

Commit:
test cases

Commit:
test cases

Commit:
test cases

Commit:
test cases

Commit:
merge issues

Commit:
removed background report results

Commit:
removed background report results

Commit:
removed background report results

Commit:
removed background report results

Commit:
removed check

Commit:
removed check

shriharishastry commented:
Have added Previous report direct download link. Also trigger button has been added at centre of the page instead of on header.
![screenshot from 2018-06-26 22-06-58](https://user-images.githubusercontent.com/10257870/41926608-7c50fae2-798d-11e8-9277-296bdc58696d.png)

Also the name of doctype has been changed to **Prepared Report**

shriharishastry committed:
frappe csv utils usage

shriharishastry committed:
frappe csv utils usage

shriharishastry committed:
frappe csv utils usage

shriharishastry committed:
frappe csv utils usage

shriharishastry committed:
frappe csv utils usage and save file

shriharishastry committed:
frappe csv utils usage and save file

shriharishastry committed:
report changes

shriharishastry committed:
report changes

Commit:
removed unused imports

Commit:
removed unused imports

saurabh6790 committed:
[Security][fix] To avoid possible sql injection via filters and or_filters parameters and tighten the field level checks (#5721)

* [fix] sanitize filters and or_filters to avoid sql injection

* add test cases for filter sanitizer

* codacy fix

* added test cases to test valid scenarios

Commit:
updated with dev of frappe

Commit:
updated with dev of frappe

pratu16x7 committed:
[init] add show report in prepared report

pratu16x7 committed:
[init] add show report in prepared report

Commit:
added check not trigger run if prepared

Commit:
added check not trigger run if prepared

Commit:
merge issues resolved

pratu16x7 committed:
[prepared-report] add download report

pratu16x7 committed:
[prepared-report] add download report

pratu16x7 committed:
[merge] conflicts

Commit:
changes on cleaning colmn headers

Commit:
changes on cleaning colmn headers

Commit:
merge conflicts

Commit:
header clean up

Commit:
header clean up

pratu16x7 committed:
[report] add prepared report filters and creation message

pratu16x7 committed:
[report] add prepared report filters and creation message

Commit:
stash issue

Commit:
stash issue

Commit:
default report for prepared

Commit:
default report for prepared

shriharishastry committed:
changes on prepared report if present

shriharishastry committed:
changes on prepared report if present

Commit:
generate prepared button report

pratu16x7 committed:
[report] make run() agnostic of data source

pratu16x7 committed:
[report] make run() agnostic of data source

pratu16x7 committed:
[report] add set_filters()

pratu16x7 committed:
[report] add set_filters()

pratu16x7 committed:
[merge] conflicts

Commit:
added checks for if no data present

Commit:
added checks for if no data present

shridarpatil committed:
ignore save password based on fieldname (#5749)

* ignore save password based on fieldname

* ignore save password if flag is True

* if ignore_password is set  then check for field

surajshetty3416 committed:
Check if permission needs to be ignored in link fields (#5747)

shridarpatil committed:
syntax error in whitelisted function throws 404 (#5743)

pratu16x7 committed:
[report] make it a prepared if query times out

pratu16x7 committed:
[report] make it a prepared if query times out

nabinhait committed:
multiple minor fixes (#5756)

pratu16x7 committed:
[fix][remove] webform section-body for alignment

pratu16x7 committed:
Merge pull request #5761 from pratu16x7/webform-fixes

[fix][remove] webform section-body for alignment

rmehta committed:
[fix] compileall to exclude node_modules

shridarpatil committed:
Address title field changed to mandatory (#5759)

netchampfaris committed:
Remove !important rules so that custom css in Print Format can override (#5762)

rmehta committed:
[fix] datepicker freezing

rmehta committed:
[minor] remember dashboard expand/collapse preference

surajshetty3416 committed:
Add logic to check if the document was modified (#5746)

- Check if the document was modified after the workflow
action mail was sent.
If yes, then show an alert with the latest document preview.

surajshetty3416 committed:
Return if no link fields found (#5764)

pratu16x7 committed:
[build-docs][context] look for _sidebar.json in parent directories (#5715)

* [build-docs][context] look for _sidebar.json in parent directories

* [build-docs] stop sidebar search at /www dir

* [website-render] use look_for_sidebar_json hook

* [website-sidebar] add hooks check

Commit:
[Bug Fix] Data import from non EN language template failed (#5760)

* Update importer.py

clear user cache to force get most updated user lang when user only changed language not login again
fix the bug of too many values to unpack

* Update importer.py

fix the bug failure to import from non EN template

frappe-pr-bot committed:
[Translation] Updated Translations (#5763)

shridarpatil committed:
removed on_trash method (#5757)

trying to update in doctype issue
fails to update if erpnext is not installed

Zlash65 committed:
[Fix] Defaults doesn't comply with user permission fix (#5735)

* added additional check for user permission
If the key for default is found in user permission, then the value is checked if its permitted to the user or not

* user permission check on defaults

* test case for defaults

codingCoffee committed:
regex fix (#5765)

courtesy: @netchanpfaris
initally the regex used to allow only digits 0-9, not it considers the
entire number range

netchampfaris commented on frappe/desk/query_report.py:
```diff
@@ -29,6 +32,90 @@ def get_report_doc(report_name):
+			# The JOB:
+			try:
+				res = frappe.get_attr(method_name)(frappe._dict(filters))
+			except:
```
Report can fail due to multiple reasons, catch only the Timeout Error

nabinhait committed:
minor fix

pratu16x7 committed:
[report] add report status

pratu16x7 committed:
[report] add report status

pratu16x7 committed:
[report-background] alerts on start and end

pratu16x7 committed:
[report-background] alerts on start and end

pratu16x7 commented:
We just have to handle a few resources.

- [ ] Periodically clear old prepared reports, maybe a scheduled job
- [ ] On trash, delete CSV files

rmehta commented:
Can you update the screenshots / gif with the latest updates?

surajshetty3416 committed:
Remove hardcoded roles for editing users (#5726)

* Remove hardcoded roles for editing users

- Allow user edit if user has role which
has access with permlevel >= 1 to user doctype

* Fix codacy

rmehta committed:
[quick-entry] show all fields if less than 7

rmehta committed:
[patch] Get apps for docs (#5772)

* [docs] add getting docs patch

* [fix] get apps for docs

pratu16x7 commented:
Updated

Commit:
[New Feature] Global Search By User Tag for every doc (#5711)

* Update global_search.py

add user tag

* Update tags.py

trigger update global search table

* Update tags.py

remove trailing space

* Update global_search.py

remove trailing space

* Update global_search.py

remove title user tag

* Update global_search.py

* Update global_search.py

rmehta committed:
[fix] frappe.safe_decode in desk.py

shriharishastry committed:
[report] remove file on trash

shriharishastry committed:
[report] remove file on trash

saurabh6790 committed:
[fix][test-case] filter sanitizer

saurabh6790 committed:
Merge pull request #5782 from saurabh6790/filter_sanitizer_test_case_fixes

[fix][test-case] filter sanitizer

pratu16x7 committed:
[formatters] add check for boot.sysdefaults, fixes frappe/erpnext#14796

pratu16x7 committed:
Merge pull request #5786 from pratu16x7/float-precision

[formatters] add check for boot.sysdefaults, fixes frappe/erpnext#14796

Zlash65 committed:
[ Fixes ] Data Import / Data Export (#5776)

* add button to toggle check all or none

* add 'M' to label for mandatory field - manual selection

* add select_all / deselect_all button to toggle checkbox

* multicheck control fix
call refresh of parent class as its overridden, ignore pushing same data in selected_options

* pass with_data parameter as 1 or 0 to be interpreted using cint

* make fields visible and disabled according to option selected

* allow editing for non mandatory field when mandatory option selected
simplified columns selection logic when passing it to be exported

felixvarghese committed:
Connect to LDAP Server using StartTLS (#5784)

* Connect to LDAP Server using StartTLS

Connect to LDAP server using StartTLS for security. An option to require trusted certificates is also included.

* fix try

* remove trailing whitespace

ioweb-gr committed:
Fix bug with smtplib and crammd5 servers that prevents email sending (#5687)

* Fix bug with smtplib and crammd5 servers that prevents email sending

* Use frappe.safe_decode instead of str

* Revert "Use frappe.safe_decode instead of str"

This reverts commit 272aa167fd86f90129517666b2cad85363026f87.

* Using frappe_encode to 'ascii' to send the password

* Update smtp.py

* Update smtp.py

saurabh6790 committed:
bumped to version 10.1.38

saurabh6790 committed:
Merge branch 'hotfix'

saurabh6790 committed:
resolved merge conflicts

rmehta committed:
[remove] mycli

codingCoffee committed:
safe encode is not required

Zlash65 committed:
make content as read only (#5787)

if message type is file, make content read only

codingCoffee committed:
Merge pull request #5792 from codingCoffee/mail_fix_123

safe encode is not required

rmehta committed:
[fix] test_db_query.py

rmehta committed:
[email] smtp connection per sender (#5791)

* [fix]email error due to incorrect smtp server auth

- A dict of smtpserver objects is created for every sender
- A smtp server object is created if not present for the given sender
- smtpserver object is picked based on the sender of the mail

* [fix] remove cache from queue.py

pratu16x7 committed:
[fix] fulfill datetime requirement for webform

pratu16x7 committed:
[global-search] add doc check for route

saurabh6790 committed:
Merge pull request #5798 from pratu16x7/hotfix

[global-search] add doc check for route

saurabh6790 committed:
bumped to version 10.1.39

saurabh6790 committed:
Merge branch 'hotfix'

saurabh6790 committed:
[fix] merge-conflicts

rmehta committed:
[style] for webforms and login

surajshetty3416 committed:
Run user-permission patch before droping apply_user_permission column (#5794)

rmehta committed:
[fix] test

adityahase committed:
Allow multiple values to be passed as arguments to rpush and lpush (#5802)

rmehta committed:
[test] [style]

rmehta committed:
[style] make page-card self contained

pratu16x7 committed:
Merge pull request #5806 from pratu16x7/website-datetime

[fix] fulfill datetime requirement for webform

saurabh6790 committed:
test case fixes (#5807)

frappe-pr-bot committed:
[Translation] Updated Translations (#5801)

Zlash65 committed:
skip property setter for naming series (#5796)

rmehta committed:
[fix] frappe.query_report_filters_by_name

sushantnadkar committed:
Convert doc field value to string type before using (#5795)

rmehta commented:
Is this complete?

netchampfaris committed:
Moment (#5808)

- Add npm packages
- Add moment builds in frappe-web
- Remove import statements to avoid duplicate loading of moment

netchampfaris committed:
Update frappe-datatable styles (#5813)

netchampfaris committed:
Create a moment bundle and include it in webform

pratu16x7 committed:
Merge pull request #5818 from netchampfaris/concat-build

Create a moment bundle and include it in webform

adityahase committed:
Set pager in mariadb console (#5809)

netchampfaris committed:
[fix] frappe.query_report should be null if report is not initialized

pratu16x7 committed:
Merge pull request #5821 from netchampfaris/fix-query-report

[fix] frappe.query_report should be null if report is not initialized

shriharishastry committed:
codacy checks resolved

shriharishastry committed:
codacy checks resolved

Commit:
codacy changes

Commit:
merge conflicts solved

Commit:
codacy issues

Commit:
codacy issues

Commit:
codacy issues
